firedrake/preconditioners/asm.py

```diff
@@ -0,0 +1,98 @@
+import abc
+
+from pyop2.mpi import COMM_SELF
+from firedrake.preconditioners.base import PCBase
+from firedrake.petsc import PETSc
+from firedrake.dmhooks import get_function_space
+
+__all__ = ("ASMPatchPC", "ASMLinesmoothPC")
+
+
+class ASMPatchPC(PCBase):
+    ''' PC for PETSc PCASM
+
+    should implement:
+    - :meth:`get_patches`
+    '''
+
+    _prefix = "pc_asm_"
+
+    def initialize(self, pc):
+        # Get context from pc
+        _, P = pc.getOperators()
+        dm = pc.getDM()
+        self.prefix = pc.getOptionsPrefix() + self._prefix
+
+        # Extract function space and mesh to obtain plex and indexing functions
+        V = get_function_space(dm)
+
+        # Obtain patches from user defined funtion
+        ises = self.get_patches(V)
+
+        # Create new PC object as ASM type and set index sets for patches
+        asmpc = PETSc.PC().create(comm=pc.comm)
+        asmpc.incrementTabLevel(1, parent=pc)
+        asmpc.setOptionsPrefix(self.prefix + "_sub")
+        asmpc.setOperators(*pc.getOperators())
+        asmpc.setType(asmpc.Type.ASM)
+        asmpc.setASMLocalSubdomains(len(ises), ises)
+        asmpc.setFromOptions()
+        self.asmpc = asmpc
+
+    @abc.abstractmethod
+    def get_patches(self, V):
+        ''' Get the patches used for PETSc PSASM
+        '''
+        pass
+
+    def update(self, pc):
+        pass
```

Comment:
This is wrong if the matrix changes. You should call `self.asmpc.setUp()`

Reply:
done